libavcodec/jpeg2000dec.c

```diff
@@ -1067,100 +1072,339 @@ static int jpeg2000_decode_packet(Jpeg2000DecoderContext *s, Jpeg2000Tile *tile,
+                        // 0 means newpasses is a RAW SigProp; 1 means newpasses is a RAW MagRef pass
+                        if (newpasses > 1)
+                            segment_passes = 2 - (cblk->npasses - bypass_term_threshold) % 3;
+                        else
+                            segment_passes = 1;
+                        bits_to_read = (uint8_t) (segment_passes - 1);
+                        next_segment_passes = 1;
+                    } else {
+                        // newpasses is an isolated Cleanup pass that precedes a RAW SigProp pass
+                        segment_passes = 1;
+                        next_segment_passes = 2;
+                    }
+                    bits_to_read = (uint8_t) (bits_to_read + cblk->lblock);
+                    segment_bytes = get_bits(s, bits_to_read);
                 }
-
-                if (newpasses > 1 && (codsty->cblk_style & JPEG2000_CTSY_HTJ2K_F)) {
-                    // Retrieve pass lengths for each pass
-                    int href_passes =  (cblk->npasses + newpasses - 1) % 3;
-                    int eb = av_log2(newpasses - href_passes);
-                    int extra_bit = newpasses > 2 ? 1 : 0;
-                    if ((ret = get_bits(s, llen + eb + 3)) < 0)
-                        return ret;
-                    cblk->pass_lengths[0] = ret;
-                    if ((ret = get_bits(s, llen + 3 + extra_bit)) < 0)
-                        return ret;
-                    cblk->pass_lengths[1] = ret;
-                    ret = cblk->pass_lengths[0] + cblk->pass_lengths[1];
+                // Update cblk->npasses and write length information
+                cblk->npasses = (uint8_t) (cblk->npasses + segment_passes);
+                segment_length[nb_segments++] = segment_bytes;
+
+                if ((cblk->modes & JPEG2000_CTSY_HTJ2K_F) && cblk->ht_plhd == HT_PLHD_OFF) {
+                    newpasses -= (uint8_t) segment_passes;
+                    primary_passes = (uint8_t) (segment_passes + cblk->fast_skip_passes);
+                    cblk->fast_skip_passes = 0;
+                    primary_bytes = segment_bytes;
+                    secondary_passes = 0;
+                    secondary_bytes = 0;
+                    empty_set = 0;
+                    if (next_segment_passes == 2 && segment_bytes == 0)
+                        empty_set = 1;
+                    while (newpasses > 0) {
+                        if (newpasses > 1)
+                            segment_passes = next_segment_passes;
+                        else
+                            segment_passes = 1;
+                        next_segment_passes = (uint8_t) (3 - next_segment_passes);
+                        bits_to_read =
+                                (uint8_t) (cblk->lblock + (unsigned int) (segment_passes) - 1);
+                        segment_bytes = get_bits(s, bits_to_read);
+                        newpasses -= (uint8_t) (segment_passes);
```

Comment:
Is this guaranteed to be > 0?

Reply:
Possible values of `segment_passes` are 0, 1, or 2 because possible values of `next_segment_passes` are 0, 1, or 2. `segment_passes == 2` only becomes true when `newpasses > 1`, namely, `newpasses >= 2`. In conclusion, yes, this is guaranteed to be `>= 0` (not `> 0`)